File: Python/02_erasing_maximum.py
import sys


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def find_argmax(array, m, count):
    max_indices = [ix for ix, item in enumerate(array) if item == m]
    if count >= 3:
        return max_indices[2]
    return max_indices[0]


def remove_maximum(array):
    m = max(array)
    count = sum(1 for item in array if item == m)
    max_ix = find_argmax(array, m, count)
    return [item for ix, item in enumerate(array) if ix != max_ix]


def main():
    reader = sys.stdin
    _ = next(reader)
    array = convert_to_intlist(next(reader))
    result = remove_maximum(array)
    print(' '.join([str(item) for item in result]))



if __name__ == '__main__':
    main()

File: Haskell/02-erasing-maximum.hs
convertToIntList :: String -> [Int]
convertToIntList = map read . words


findArgmax :: [Int] -> Int -> Int -> Int
findArgmax array maxValue count =
    let maxIndices = snd $ unzip $ filter (\(x, _) -> x == maxValue) $ zip array [0..]
    in if count >= 3 then maxIndices !! 2 else head maxIndices


removeMaximum :: [Int] -> [Int]
removeMaximum array =
    let maxValue = maximum array
        count = foldl (\acc x -> if x == maxValue then acc + 1 else acc) 0 array
        maxIx = findArgmax array maxValue count
    in fst $ unzip $ filter (\(_, ix) -> ix /= maxIx) $ zip array [0..]


main :: IO()
main = do
    _ <- readLn :: IO Int
    array <- fmap convertToIntList getLine
    let result = removeMaximum array
    putStrLn $ unwords $ map show result
